### PR TITLE
replace from/to server versions in layouts

### PR DESCRIPTION
**Pull request by GuyAfik**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-5195)

## Description
when using `demisto-sdk upload` or `demisto-sdk prepare-content` make sure the `fromversion` and `toversion` of layouts will be transferred into `fromServerVersion` and `toServerVersion`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 * Fixed an issue where the **upload** command failed because `docker_native_image_config.json` file could not be found.
 * Added a metadata file to the content graph zip, to be used in the **update-content-graph** command.
 * Updated the **validate** and **update-release-notes** commands to unskip the *Triggers Recommendations* content type.
+* Fixed an issue where the **upload** and **prepare-content** commands didn't add `fromServerVersion` and `toServerVersion` to layouts.
```
Comment (dorschw) [inline CHANGELOG.md:12]: please move to `Unreleased`

Comment (GuyAfik) [reply CHANGELOG.md]: you are right, good catch

File: demisto_sdk/commands/upload/tests/uploader_test.py [hunk 2]
```diff
@@ -206,19 +206,60 @@ def test_upload_layout_positive(demisto_client_configure, mocker):
...
+    unify_mocker = mocker.patch.object(LayoutObject, "_unify")
+
     layout_name = "layout-details-test_bla-V2.json"
     layout_path = f"{git_path()}/demisto_sdk/tests/test_files/Packs/DummyPack/Layouts/{layout_name}"
+
     uploader = Uploader(input=layout_path, insecure=False, verbose=False)
     mocker.patch.object(uploader, "client")
     uploader.upload()
 
+    assert not unify_mocker.called
```
Comment (dorschw) [inline demisto_sdk/commands/upload/tests/uploader_test.py:225]: just to make sure, is `.called` a property or a method?

File: demisto_sdk/tests/integration_tests/unify_integration_test.py [hunk 3]
```diff
@@ -283,3 +284,47 @@ def test_ignore_native_image_script(self, repo):
...
+            )
+            assert not result.exception
+            assert not result.stderr
+            with open(os.path.join(output)) as updated_layout:
+                layout_data = json.load(updated_layout)
+                assert "fromVersion" in layout_data
+                assert "fromServerVersion" in layout_data
+                assert "toServerVersion" in layout_data
+
+                assert layout_data["fromVersion"] == layout_data["fromServerVersion"]
```
Comment (dorschw) [inline demisto_sdk/tests/integration_tests/unify_integration_test.py:330]:
```suggestion
                assert layout_data["fromVersion"] == layout_data["fromServerVersion"]
                assert layout_data["toVersion"] == layout_data["toServerVersion"]

```

File: demisto_sdk/tests/integration_tests/unify_integration_test.py [hunk 4]
```diff
@@ -283,3 +284,47 @@ def test_ignore_native_image_script(self, repo):
...
+        output = "test.json"
+
+        with ChangeCWD(pack.repo_path):
+            runner = CliRunner(mix_stderr=False)
+            result = runner.invoke(
+                main, [UNIFY_CMD, "-i", f"{layout.path}", "-o", output]
+            )
+            assert not result.exception
+            assert not result.stderr
+            with open(os.path.join(output)) as updated_layout:
```
Comment (dorschw) [inline demisto_sdk/tests/integration_tests/unify_integration_test.py:324]: Pathlib is preferred over os.path

File: demisto_sdk/tests/integration_tests/unify_integration_test.py (hunk 3 above)
Comment (dorschw) [inline demisto_sdk/tests/integration_tests/unify_integration_test.py:330]: all these can be one indentation level back

File: demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py [hunk 5]
```diff
@@ -133,3 +136,26 @@ def is_file_structure_list(self) -> bool:
...
+
+        Args:
+            dest_dir: Destination directory.
+            output: output suffix to add the destination directory.
+
+        Returns:
+            List[Path]: List of new created unified json files.
+        """
+        if isinstance(dest_dir, str):
+            dest_dir = Path(dest_dir)
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:151]: yes, empty string makes the path default to the current working directory, however need to make sure its not `None`

File: demisto_sdk/commands/upload/tests/uploader_test.py (hunk 2 above)
Comment (GuyAfik) [reply demisto_sdk/commands/upload/tests/uploader_test.py:225]: propery

File: demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py [hunk 6]
```diff
@@ -133,3 +136,26 @@ def is_file_structure_list(self) -> bool:
...
+        # Unify step
+        return [
+            Path(
+                str(
+                    PrepareUploadManager.prepare_for_upload(
+                        input=self.path, output=dest_dir / output  # type: ignore[operator]
```
[the remark this answers is not in the document]
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:157]: agreed

[the remark this answers is not in the document]
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:157]: however it didn't really help eventually
```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:160: error:
Argument 1 to "Path" has incompatible type "Union[Path, str]"; expected "str"
[arg-type]
                            input=self.path, output=Path(dest_dir, output)
```

[the remark this answers is not in the document]
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:157]: but it works with both Path and str

File: demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py [hunk 7]
```diff
@@ -133,3 +136,29 @@ def is_file_structure_list(self) -> bool:
...
+        """
+        if dest_dir is None:
+            dest_dir = ""
+
+        # Unify step
+        return [
+            Path(
+                str(
+                    PrepareUploadManager.prepare_for_upload(
+                        input=self.path, output=Path(dest_dir, output)  # type: ignore[arg-type]
```
Comment (dorschw) [inline demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:160]: is the `type: ignore[...]` necessary if we do `Path`? 
Also, please manually check to make sure it works with both a string and `Path`

**Review by dorschw — APPROVED** [review]

File: demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py (hunk 7 above)
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:160]: see my comment here: https://github.com/demisto/demisto-sdk/pull/2679#discussion_r1067840722

**Review by ilaner — APPROVED** [review]
Great job

File: demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py (hunk 7 above)
Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:160]: its necessary even when we do it with path, however both string and path work

Comment (GuyAfik) [reply demisto_sdk/commands/common/content/objects/pack_objects/abstract_pack_objects/json_content_object.py:160]: checked it manually and its ok

**Pull request merged**